elizaOS/eliza
# feat: Add Cloudflare AI Gateway support

# Cloudflare AI Gateway Integration

## Relates to
Enhances model provider integration by adding support for Cloudflare AI Gateway

![image](https://github.com/user-attachments/assets/838a899b-a252-496e-8f29-c2fc70900665)


## Risks
Low - Changes are isolated to model provider routing and include fallback mechanisms:
- If gateway configuration is incomplete/invalid, falls back to direct API calls
- Comprehensive logging added for debugging
- No changes to core model interaction logic

## Background

### What does this PR do?
Adds support for routing API calls through Cloudflare AI Gateway, providing:
- Improved latency through Cloudflare's global network
- Detailed analytics and monitoring
- Cost optimization through request caching
- Enhanced security and access control

### Changes Made

1. `packages/core/src/generation.ts`:
   - Added `getCloudflareGatewayBaseURL` utility function for consistent gateway URL construction
   - Implemented gateway support for OpenAI, Anthropic, and Groq providers
   - Added comprehensive logging for debugging
   - Implemented fallback logic when gateway is disabled/misconfigured

2. `docs/docs/guides/configuration.md`:
   - Added new section documenting Cloudflare AI Gateway integration
   - Included configuration instructions and environment variables
   - Listed supported providers and benefits
   - Explained URL format and fallback behavior

3. `.env.example`:
   - Added Cloudflare Gateway configuration variables:
     ```bash
     CLOUDFLARE_GW_ENABLED=true
     CLOUDFLARE_AI_ACCOUNT_ID=your-account-id
     CLOUDFLARE_AI_GATEWAY_ID=your-gateway-id
     ```

### What kind of change is this?
Feature (non-breaking change which adds functionality)

## Documentation changes
Documentation has been updated in `configuration.md` with:
- Gateway setup instructions
- Configuration options
- Benefits and supported providers

## Testing

### Where should a reviewer start?
1. Review the `getCloudflareGatewayBaseURL` function in `generation.ts`
2. Check gateway integration in provider handlers
3. Verify documentation clarity in `configuration.md`

### Detailed testing steps
1. Configure Cloudflare AI Gateway:
   ```bash
   CLOUDFLARE_GW_ENABLED=true
   CLOUDFLARE_AI_ACCOUNT_ID=your-account-id
   CLOUDFLARE_AI_GATEWAY_ID=your-gateway-id
   ```

2. Test with different providers:
   - Verify OpenAI requests route through gateway
   - Verify Anthropic requests route through gateway
   - Verify Groq requests route through gateway

3. Test fallback behavior:
   - Disable gateway and verify direct API calls
   - Use invalid configuration and verify fallback
   - Check logging messages for debugging info

### Expected Results
- When gateway is properly configured:
  - Requests route through Cloudflare
  - Logs show gateway URL being used
  - Response handling remains unchanged

- When gateway is disabled/misconfigured:
  - Fallback to direct API calls
  - Warning logs indicate configuration issues
  - Service continues to function normally

## Discord username
w3_bounty
